### PR TITLE
Gaurav route

### PR DESCRIPTION
**Pull request by gauravThakur-sgl**
#### liveurl - https://timelogone.vercel.app

### CODE REVIEW

File: gaurav/package.json [hunk 1]
```diff
@@ -10,21 +10,28 @@
...
-
-    "react-router-dom": "^7.0.1",
-    "serve": "^14.2.4",
-    "vite-plugin-pwa": "^0.21.0",
-
     "react-redux": "^9.1.2",
-    "redux-persist": "^6.0.0"
+    "react-router-dom": "^7.0.2",
+    "redux-persist": "^6.0.0",
+    "serve": "^14.2.4",
```
Comment (shubhamships) [inline gaurav/package.json:25]: what's the use of `serve`?

File: gaurav/package.json [hunk 2]
```diff
@@ -10,21 +10,28 @@
...
+    "shadcn": "^2.1.6",
```
Comment (shubhamships) [inline gaurav/package.json:26]: this is not needed to be installed as a package

File: gaurav/src/App.tsx [hunk 3]
```diff
@@ -1,9 +1,29 @@
...
+  const router = createBrowserRouter([
+    {
+      path: "/",
+      element: <Root />,
+      children: [
+        {
+          path: "timelog",
+          element: <MarkTime />,
+        },
+        {
```
Comment (shubhamships) [inline gaurav/src/App.tsx:15]: no point of creating nested routes if we have just two component files

File: gaurav/src/components/ButtonUI.tsx [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import Button from "./ui/Button";
+import Button from "./ui/personal/Button";
```
Comment (shubhamships) [inline gaurav/src/components/ButtonUI.tsx:1]: code duplication!!!

File: gaurav/src/components/Log.tsx [hunk 5]
```diff
@@ -0,0 +1,160 @@
+import Button from "./ui/personal/Button";
+import Card from "./ui/Card";
+import { useState, useEffect } from "react";
+import Errors from "./ui/Errors";
+import { useLocation } from "react-router-dom";
+// import { useLocation } from "react-router-dom";
```
Comment (shubhamships) [inline gaurav/src/components/Log.tsx:6]: clean up this file and organize code for readability

File: gaurav/src/components/MarkTime.tsx [hunk 6]
```diff
@@ -45,35 +51,53 @@ export const MarkTime = () => {
...
-      const newNetTime = (punchOutTime - punchInTime) / 1000 + netTime;
-      setNetTime(newNetTime); // time in seconds
-      localStorage.setItem("netTime", JSON.stringify(newNetTime));
+    if ((punchOutTime - punchInTime) / 1000 < 120) {
+      setIsError(true);
+    } else {
+      console.log("success");
+      if (punchData.length === 0 || punchData[punchData.length - 1].type === "IN") {
+        const newPunchOutData = [
+          ...punchData,
```
Comment (shubhamships) [inline gaurav/src/components/MarkTime.tsx:60]: reuse already written logic using function modularity instead of copy and pasting... code duplication is bad

File: gaurav/src/index.css [hunk 7]
```diff
@@ -6,5 +6,116 @@
...
+
+        --foreground: 0 0% 3.9%;
+
+        --card: 0 0% 100%;
+
+        --card-foreground: 0 0% 3.9%;
+
+        --popover: 0 0% 100%;
+
+        --popover-foreground: 0 0% 3.9%;
```
Comment (shubhamships) [inline gaurav/src/index.css:21]: remove extra lines that are empty

File: gaurav/src/routes/Root.tsx [hunk 8]
```diff
@@ -1,19 +1,40 @@
...
-      <div className="flex justify-center items-center bg-purple-500">
-        <ul className="flex justify-center items-center gap-3 p-3">
-          <li className="text-white font-medium">
-            <Link to="/">Home</Link>
-          </li>
-          <li className="text-white font-medium">
-            <Link to="/login">Login</Link>
-          </li>
+      <div className="flex justify-center items-center fixed bottom-0 w-full">
+        <ul className="flex justify-between md:justify-center items-center bg-primary gap-6 p-2 px-4 w-full">
```
Comment (shubhamships) [inline gaurav/src/routes/Root.tsx:7]: Menus can be improved.... it should have more padding and better UI. Also add some content on home route

File: gaurav/tailwind.config.js [hunk 9]
```diff
@@ -1,19 +1,67 @@
...
+				// 	DEFAULT: '#1f499e',
+				// 	foreground: 'hsl(var(--primary-foreground))'
+				// },
+				primary: '#1f499e',
+				destructive: {
+					DEFAULT: 'hsl(var(--destructive))',
+					foreground: 'hsl(var(--destructive-foreground))'
+				},
+				success: '#16a34a',
+				soft: '#e0e7ff',
```
Comment (shubhamships) [inline gaurav/tailwind.config.js:20]: Fix indentation and spacing

File: gaurav/tsconfig.json [hunk 10]
```diff
@@ -5,7 +5,7 @@
     "lib": ["ES2020", "DOM", "DOM.Iterable"],
     "module": "ESNext",
     "skipLibCheck": true,
-
+    "composite": true,
```
Comment (shubhamships) [inline gaurav/tsconfig.json:8]: Why was this added? Please explain here.

File: gaurav/tsconfig.json [hunk 11]
```diff
@@ -21,11 +21,12 @@
     "noFallthroughCasesInSwitch": true,
 
     /* directory path modifier */ // not working with new vite installation
-    "baseUrl": "",
+    "baseUrl": ".",
     "paths": {
-      "@/*": ["src/*"]
+      "@/*": ["./src/*"],
```
Comment (shubhamships) [inline gaurav/tsconfig.json:26]: this seems hacky... kindly fix the existing issue instead of workarounds

File: gaurav/vite.config.ts [hunk 12]
```diff
@@ -14,6 +14,7 @@ export default defineConfig({
   resolve: {
     alias: {
       "@": path.resolve(__dirname, "./src"),
+      '@shadcn': '/node_modules/@shadcn',
```
Comment (shubhamships) [inline gaurav/vite.config.ts:17]: Why this? its not even being used anywhere

File: gaurav/src/components/ButtonUI.tsx (hunk 4 above)
Comment (gauravThakur-sgl) [reply gaurav/src/components/ButtonUI.tsx:1]: One button is from `shadcn` which is used in the calander,
and the other one is custom made button Commponent.

File: gaurav/package.json (hunk 1 above)
Comment (gauravThakur-sgl) [reply gaurav/package.json]: `serve` is used to setting up local servers to host static web projects that are in development.
Removed as it is remained unused

File: gaurav/package.json (hunk 2 above)
Comment (gauravThakur-sgl) [reply gaurav/package.json:25]: removed both `serve` and `shadcn`

File: gaurav/src/App.tsx (hunk 3 above)
Comment (gauravThakur-sgl) [reply gaurav/src/App.tsx]: Removed nested Routes.

File: gaurav/src/components/Log.tsx (hunk 5 above)
Comment (gauravThakur-sgl) [reply gaurav/src/components/Log.tsx]: Done

File: gaurav/src/index.css (hunk 7 above)
Comment (gauravThakur-sgl) [reply gaurav/src/index.css:21]: Removed extra space from the `index.css` file

File: gaurav/tailwind.config.js (hunk 9 above)
Comment (gauravThakur-sgl) [reply gaurav/tailwind.config.js:20]: Fixed indentation and spacing

File: gaurav/tsconfig.json (hunk 10 above)
Comment (gauravThakur-sgl) [reply gaurav/tsconfig.json:8]: removed `"composite": true` as it is not used anywhere in the project

File: gaurav/vite.config.ts (hunk 12 above)
Comment (gauravThakur-sgl) [reply gaurav/vite.config.ts:17]: Removed

File: gaurav/src/components/MarkTime.tsx (hunk 6 above)
Comment (gauravThakur-sgl) [reply gaurav/src/components/MarkTime.tsx:60]: created `lastPunchData = punchData[punchData.length - 1]`
and used it int the place of  `punchData[punchData.length - 1].type === "IN"` i.e `lastPunchData.type === "IN`

**Comment by gauravThakur-sgl** [comment]
@shubhamships  Sir, changes done according to the comments

File: gaurav/src/components/MarkTime.tsx [hunk 13]
```diff
@@ -41,39 +52,60 @@ export const MarkTime = () => {
...
+      if (punchData.length === 0 || lastPunchData.type === "IN") {
+        const newPunchOutData = [
+          ...punchData,
+          {
+            id: punchData.length,
+            time: new Date().toISOString(),
+            type: "OUT",
+          },
+        ];
+        setPunchData(newPunchOutData);
```
Comment (shubhamships) [inline gaurav/src/components/MarkTime.tsx:73]: please clean up this file too @gauravThakur-sgl

File: gaurav/package.json [hunk 14]
```diff
@@ -10,21 +10,26 @@
...
-    "react-router-dom": "^7.0.1",
-    "serve": "^14.2.4",
-    "vite-plugin-pwa": "^0.21.0",
-
     "react-redux": "^9.1.2",
-    "redux-persist": "^6.0.0"
+    "react-router-dom": "^7.0.2",
+    "redux-persist": "^6.0.0",
+    "tailwind-merge": "^2.5.5",
+    "tailwindcss-animate": "^1.0.7",
```
Comment (shubhamships) [inline gaurav/package.json:26]: are we using this? if yes, where?

File: gaurav/src/App.tsx [hunk 15]
```diff
@@ -1,9 +1,52 @@
...
-  </div>;
+  return (
+    <>
+      <BrowserRouter>
+        <div className="flex justify-center items-center fixed bottom-0 w-full border">
+          <ul className="flex justify-between md:justify-center items-center bg-white gap-6 p-2 px-8 w-full">
+            <MenuLink name="Home" route="/" iconName={<House className="w-5 h-5 text-gray-500" />} />
+            <MenuLink name="Log" route="login" iconName={<Calendar className="w-5 h-5 text-gray-500" />} />
+            <MenuLink name="Timelog" route="timelog" iconName={<Clock className="w-5 h-5 text-gray-500" />} />
+          </ul>
```
Comment (ChinmaySingh-ShipGlobal) [inline gaurav/src/App.tsx:15]: You can reuse the className by declaring a variable as it is being repeated here for icons.

File: gaurav/src/components/Log.tsx [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+  });
+
+  // Fetching data from local storage
+  useEffect(() => {
+    const punchData = JSON.parse(localStorage.getItem("punchData") || "[]");
+    const newTotalTime = JSON.parse(localStorage.getItem("totalTime") || "[]");
+    setPunchData(punchData);
+    setTotalTime(newTotalTime);
+  }, []);
+  console.log(punchData, "punchDatahaa");
```
Comment (ChinmaySingh-ShipGlobal) [inline gaurav/src/components/Log.tsx:30]: Remove console logs where ever not necessary.

File: gaurav/src/components/MarkTime.tsx [hunk 17]
```diff
@@ -1,16 +1,24 @@
...
 export const MarkTime = () => {
+  const location = useLocation();
+  const queryParams = new URLSearchParams(location.search);
+  const selectedDate = new Date(queryParams.get("date") || new Date().toISOString().split("T")[0]);
+
   const [netTime, setNetTime] = useState<number>(0);
-  const [isCompleted, setIsCompleted] = useState(false);
+  const [totalTime, setTotalTime] = useState<{ time: number; date: string }[]>([]);
+  const [error, setIsError] = useState(false);
   const [punchInDisabled, setPunchInDisabled] = useState<boolean>();
```
Comment (ChinmaySingh-ShipGlobal) [inline gaurav/src/components/MarkTime.tsx:14]: Try to initialize the boolean state variables with either true or false.

File: gaurav/src/components/MarkTime.tsx [hunk 18]
```diff
@@ -1,16 +1,24 @@
...
   const [punchOutDisabled, setPunchOutDisabled] = useState<boolean>();
```
Comment (ChinmaySingh-ShipGlobal) [inline gaurav/src/components/MarkTime.tsx:15]: Try to initialize the boolean state variables with either true or false.

File: gaurav/src/components/MarkTime.tsx [hunk 19]
```diff
@@ -41,39 +52,60 @@ export const MarkTime = () => {
...
+            time: newNetTime,
+            date: new Date().toDateString(),
+          },
+        ];
+        setTotalTime(newTotalTime);
+        localStorage.setItem("totalTime", JSON.stringify(newTotalTime));
+        setIsError(false);
+      }
+      setPunchInDisabled(false);
+      setPunchOutDisabled(true);
```
Comment (ChinmaySingh-ShipGlobal) [inline gaurav/src/components/MarkTime.tsx:90]: Check if setPunchInDisabled and setPunchOutDisabled are complementary. If yes, then try to use single variable.

File: gaurav/package.json (hunk 14 above)
Comment (gauravThakur-sgl) [reply gaurav/package.json]: @shubhamships Sir,
 `react-router-dom` using for the routing,
`redux-persist`, `tailwind-merge` and `tailwind-animate` is not being used anywhere so removed.

**Comment by gauravThakur-sgl** [comment]
@shubhamships  Sir, @ChinmaySingh-ShipGlobal  Sir
fixed above issues
